tools/register_users_from_csv.py

```diff
@@ -0,0 +1,138 @@
+            "Content-Type": "application/json",
+        }
+        session.headers.update(headers)
+        return session
+
+    def create_user(self, username, first_name, last_name, email):
+        """Create a user."""
+        url = f"{self.scim_url}/Users/{username}"
+        r = self.session.get(url)
+        if r.status_code == 200:
+            print(f"User {username} exists.")
+        if r.status_code == 404:
+            print(f"Creating user {username}.")
+            payload = {
+                "schemas": ["urn:ietf:params:scim:schemas:core:2.0:User"],
+                "userName": username,
+                "name": {
+                    "givenName": first_name,
+                    "familyName": last_name,
+                },
+                "emails": [
+                    {
+                        "value": email,
+                        "type": "work",
+                        "primary": True,
+                    }
+                ]
+            }
+            r = self.session.post(
+                f"{self.scim_url}/Users",
+                json=payload
+            )
+            if r.status_code == 201:
+                print(f"Created user {username}.")
+
+    def add_user_to_group(self, username, allocation):
+        r = self.session.get(
+            f"{self.scim_url}/Groups/{allocation}"
+        )
+        if r.status_code == 200 and username not in [
+            x.get("value") for x in r.json().get("members", {})
+        ]:
+            payload = {
+                "schemas": [
+                    "urn:ietf:params:scim:api:messages:2.0:PatchOp"
+                ],
+                "Operations": [
+                    {
+                        "op": "add",
+                        "value": {
+                            "members": [
+                                {
+                                    "value": username
+                                }
+                            ]
+                        }
+                    }
+                ]
+            }
+            r = self.session.patch(
+                f"{self.scim_url}/Groups/{allocation}",
+                json=payload
+            )
+            if r.status_code in [200, 201]:
+                print(f"Added user {username} to allocation {allocation}.")
+
+
+def main():
+    client = ScimClient(
+        sys.argv[2],
+        "https://keycloak.mss.mghpcc.org",
+        os.environ.get('CLIENT_ID'),
+        os.environ.get('CLIENT_SECRET')
+    )
+
+    with open(sys.argv[1], newline='') as csvfile:
+        reader = csv.reader(csvfile, delimiter=',', quotechar='|')
+        for row in reader:
+            print(f"Processing: {row}.")
+
+            client.create_user(username=row[0],
+                               first_name=row[1],
+                               last_name=row[2],
+                               email=row[0])
+            client.add_user_to_group(username=row[0],
+                                     allocation=row[3])
```

Comment:
Good question. As I discovered today, no. I will add that in in a future update. This doesn't prevent them from logging in to OpenShift/OpenStack though. They'll just appear in the allocation listing of users, rather than the project listing of users.

Reply:
@knikolla can you create an issue for this and then mark this conversation as resolved?

Reply:
There already is #16